libs/gl-client/src/pairing/attestation_device.rs

```diff
@@ -0,0 +1,265 @@
+use std::time::{SystemTime, UNIX_EPOCH};
+
+use super::{into_approve_pairing_error, into_verify_pairing_data_error, Error};
+use crate::{
+    credentials::{RuneProvider, TlsConfigProvider},
+    pb::scheduler::{
+        pairing_client::PairingClient, ApprovePairingRequest, Empty, GetPairingDataRequest,
+        GetPairingDataResponse,
+    },
+    tls::TlsConfig,
+};
+use bytes::BufMut as _;
+use picky::{pem::Pem, x509::Csr};
+use picky_asn1_x509::{PublicKey, SubjectPublicKeyInfo};
+use ring::{
+    rand,
+    signature::{self, EcdsaKeyPair, KeyPair},
+};
+
+use rustls_pemfile as pemfile;
+use tonic::transport::Channel;
+
+type Result<T, E = super::Error> = core::result::Result<T, E>;
+
+pub struct Connected(PairingClient<Channel>);
+pub struct Unconnected();
+
+pub struct Client<T> {
+    inner: T,
+    tls: TlsConfig,
+    uri: String,
+    rune: String,
+    key: Vec<u8>,
+}
+
+impl Client<Unconnected> {
+    pub fn new<C>(creds: C) -> Result<Client<Unconnected>>
+    where
+        C: TlsConfigProvider + RuneProvider,
+    {
+        let tls = creds.tls_config();
+        let rune = creds.rune();
+        let key = tls
+            .clone()
+            .private_key
+            .ok_or(Error::BuildClientError("empty tls private key".to_string()))?;
+        Ok(Self {
+            inner: Unconnected(),
+            tls,
+            uri: crate::utils::scheduler_uri(),
+            rune,
+            key,
+        })
+    }
+
+    pub fn with_uri(mut self, uri: String) -> Client<Unconnected> {
+        self.uri = uri;
+        self
+    }
+
+    pub async fn connect(self) -> Result<Client<Connected>> {
+        let channel = tonic::transport::Endpoint::from_shared(self.uri.clone())?
+            .tls_config(self.tls.inner.clone())?
+            .tcp_keepalive(Some(crate::TCP_KEEPALIVE))
+            .http2_keep_alive_interval(crate::TCP_KEEPALIVE)
+            .keep_alive_timeout(crate::TCP_KEEPALIVE_TIMEOUT)
+            .keep_alive_while_idle(true)
+            .connect_lazy();
+
+        let inner = PairingClient::new(channel);
+
+        Ok(Client {
+            inner: Connected(inner),
+            tls: self.tls,
+            uri: self.uri,
+            rune: self.rune,
+            key: self.key,
+        })
+    }
+}
+
+impl Client<Connected> {
+    pub async fn get_pairing_data(&self, session_id: &str) -> Result<GetPairingDataResponse> {
+        Ok(self
+            .inner
+            .0
+            .clone()
+            .get_pairing_data(GetPairingDataRequest {
+                session_id: session_id.to_string(),
+            })
+            .await?
+            .into_inner())
+    }
+
+    pub async fn approve_pairing(
+        &self,
+        session_id: &str,
+        node_id: &[u8],
+        device_name: &str,
```

Comment:
Can we make the `device_name` optional?

Reply:
To make the `device_name` optional we need to think about what would make for a good identifier such that we can revoke a single rune that we handed out to a device.